### PR TITLE
SAMZA-2501 : Optimizing startpoint manager to not make successive bootstrapMessage calls to coordinator-store

### PR DESCRIPTION
**Pull request by rmatharu**
Symptom: Currently the startpoint manager queries metastore for task-ssps and for each ssp queries coordinator-store, issuing a re-read on the coordinator store. This causes increased AM startup time and depending on the number of input SSPs to a job (thousands in case of regex), YARN may timeout the AM.
Cause: Above.
Fix: This change reduces to number of metastore reads to 2; one for startpoints keyed by ssp and the other for startpoints keyed by ssp+taskname.
API changes: None
Upgrade Instructions: None

### CODE REVIEW

File: samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java [hunk 1]
```diff
@@ -158,8 +158,25 @@ public void writeStartpoint(SystemStreamPartition ssp, TaskName taskName, Startp
    * @return {@link Optional} of {@link Startpoint} for the {@link SystemStreamPartition}.
    *         It is empty if it does not exist or if it is too stale.
    */
+  @VisibleForTesting
   public Optional<Startpoint> readStartpoint(SystemStreamPartition ssp) {
-    return readStartpoint(ssp, null);
```
Comment (Sanil15) [inline samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java:162]: Revert the change on line line 162 calling `readStartpoint(ssp, null);` will still do the intended, do not need line 163-165

File: samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java [hunk 2]
```diff
@@ -158,8 +158,25 @@ public void writeStartpoint(SystemStreamPartition ssp, TaskName taskName, Startp
...
+   * Returns the last {@link Startpoint} that defines the start position for a {@link SystemStreamPartition} and {@link TaskName}.
+   * @param ssp The {@link SystemStreamPartition} to fetch the {@link Startpoint} for.
+   * @param taskName the {@link TaskName} to fetch the {@link Startpoint} for.
+   * @return {@link Optional} of {@link Startpoint} for the {@link SystemStreamPartition}.
+   *         It is empty if it does not exist or if it is too stale.
+   */
+  @VisibleForTesting
+  public Optional<Startpoint> readStartpoint(SystemStreamPartition ssp, TaskName taskName) {
+    Map<String, byte[]> startpointBytes = readWriteStore.all();
+    // there is no task-name to use as key for the startpoint in this case (only the ssp), so we use a null task-name
```
Comment (Sanil15) [inline samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java:178]: Please correct the comment, task-name can non-null

**Comment by rmatharu** [comment]
+1, simplified and addressed both comments.

File: samza-core/src/main/java/org/apache/samza/config/ClusterManagerConfig.java [hunk 3]
```diff
@@ -127,6 +127,7 @@
    */
   private static final String AM_JMX_ENABLED = "yarn.am.jmx.enabled";
   private static final String CLUSTER_MANAGER_JMX_ENABLED = "cluster-manager.jobcoordinator.jmx.enabled";
+  private static final String CLUSTER_MANAGER_STARTPOINT_FANOUT_ENABLED = "job.startpoint.fanout.enabled";
```
Comment (Sanil15) [inline samza-core/src/main/java/org/apache/samza/config/ClusterManagerConfig.java:130]: If this config is called `job.startpoint.fanout.enabled`, isn't it better to be a part of JobConfig than ClusterManagerConfig?

File: samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java (hunk 1 above)
Comment (rmatharu) [reply samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java:162]: Is used in numerous tests, so need it.

File: samza-core/src/main/java/org/apache/samza/config/ClusterManagerConfig.java (hunk 3 above)
Comment (Sanil15) [reply samza-core/src/main/java/org/apache/samza/config/ClusterManagerConfig.java:130]: Once you decide where it fits, since this is user-facing, please add docs to doc markdown. Thanks!

File: samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java (hunk 1 above)
Comment (Sanil15) [reply samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java:162]: You are not changing the function signature, so the test should still work when you invoke `readStartpoint(ssp, null)` from here what am I missing?

Comment (rmatharu) [reply samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java:162]: The fact that null as task-name maps to "only-ssp" is an internal of the class, i.e., the unit. 
Unit-tests test the unit not internal nitty gritties of the unit.
